parser_test.go

```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
+	"bytes"
+	"testing"
+	"time"
+)
+
+var (
+	timeout  = make(chan bool, 1)
+	gcChan   = make(chan *gctrace, 1)
+	scvgChan = make(chan *scvgtrace, 1)
+)
+
+func timeoutAfter(d time.Duration, t *testing.T) {
+	time.Sleep(d)
+	timeout <- true
+}
+
+func runParserWith(line string) {
+	reader := bytes.NewReader([]byte(line))
+
+	parser := Parser{
+		reader:   reader,
+		gcChan:   gcChan,
+		scvgChan: scvgChan,
+	}
+
+	parser.Run()
+}
+
+func TestParserWithMatchingInput(t *testing.T) {
+	go timeoutAfter(100*time.Millisecond, t)
+
+	line := "gc76(1): 2+1+1390+1 us, 1 -> 3 MB, 16397 (1015746-999349) objects, 1436/1/0 sweeps, 0(0) handoff, 0(0) steal, 0/0/0 yields\n"
+	ended := make(chan bool, 1)
+
+	go func() {
+		runParserWith(line)
+		ended <- true
+	}()
+
+	expectedHeapSize := 3
+
+	select {
+	case gctrace := <-gcChan:
+		if gctrace.Heap1 != expectedHeapSize {
+			t.Errorf("Expected gctrace.Heap1 to equal %d. Got %d instead.", expectedHeapSize, gctrace.Heap1)
+		}
+	case <-ended:
```

Comment:
I'm having a weird case where this channel is called before gcChan gets called... A concurrency issue that I wouldn't have expected.

I'm removing this `ended` channel and will expect a timeout in case of failure instead.